app/controllers/api/v1/study_files_controller.rb

```diff
@@ -291,60 +301,96 @@ def perform_update(study_file)
+            logger.info "Updating cluster #{@cluster.name} to match #{safe_file_params[:name]}"
+            # before updating, check if the defaults also need to change
+            if @study.default_cluster == @cluster
+              @study.default_options[:cluster] = study_file.name
+              @study.save
             end
+            @cluster.update(name: study_file.name)
           end
+        end
 
-          if ['Expression Matrix', 'MM Coordinate Matrix'].include?(safe_file_params[:file_type]) && !safe_file_params[:y_axis_label].blank?
-            # if user is supplying an expression axis label, update default options hash
-            @study.update(default_options: @study.default_options.merge(expression_label: safe_file_params[:y_axis_label]))
-          end
+        if ['Expression Matrix', 'MM Coordinate Matrix'].include?(safe_file_params[:file_type]) && !safe_file_params[:y_axis_label].blank?
+          # if user is supplying an expression axis label, update default options hash
+          options = @study.default_options.dup
+          options.merge!(expression_label: safe_file_params[:y_axis_label])
+          @study.update(default_options: options)
+        end
 
-          if parse_on_upload &&
-            study_file.parseable? &&
-            !study_file.parsing? &&
-            study_file.parse_status == 'unparsed' &&
-            study_file.status == 'uploaded' &&
-            safe_file_params[:upload].present?
-            FileParseService.run_parse_job(@study_file, @study, current_api_user)
-          end
+        if safe_file_params[:upload].present? && !is_chunked
+          complete_upload_process(study_file, parse_on_upload)
+        end
+      end
 
-          return true
+
+      # handles adding an additional chunk to a file upload.  Note that create/update must be
+      # called first with the first chunk of the file, in order to initialize the process
+      # if this is the last chunk, it will handle sending to firecloud and launching a
+      # parse job if requested
+      def chunk
+        safe_file_params = study_file_params
+        upload = safe_file_params[:upload]
+        content_range = RequestUtils.parse_content_range_header(request.headers)
+        if content_range.present?
+          if content_range[:first_byte] == 0
+            render json: {errors: 'create/update should be used for uploading the first chunk'}, status: :unprocessable_entity
+          end
+          if content_range[:first_byte] != @study_file.upload_file_size
+            render json: {errors: "Incorrect chunk sent: expected bytes starting with #{@study_file.upload_file_size}, received #{content_range[:first_byte]}" }, status: :unprocessable_entity
+          end
         else
-          return false
+          render json: {errors: 'Missing Content-Range header'}, status: :unprocessable_entity
+        end
+
+        File.open(@study_file.upload.path, "ab") do |f|
+          f.write upload.read
+        end
```

Comment:
Out of scope, but while we're here: any idea how much it'd speed upload user journey to POST chunks from this server to upstream (whether directly to the bucket or a FireCloud proxy)?

As is, we essentially upload each file twice, serially -- once from web browser to SCP server, then again from SCP server to upstream server.  The latter only _begins_ once the former ends.  "Streaming" upload from SCP to upstream (still uploading twice, but in parallel) should effectively eliminate the time for that 2nd upload phase.

(Streaming from the web browser to upstream server would be even more optimal, but maybe the effort/value ratio wouldn't be quite as worthwhile as streaming SCP server to upstream server.)